src/publish.rs

```diff
@@ -43,8 +43,8 @@ pub fn publish<T: Backend + ?Sized>(name: &str, backend: &T, env: Option<&str>)
     info!("Publishing {}={} to {}",
           name,
           version,
-          env.unwrap_or("global"));
-    backend.upload_artifact_dir(name, version, env)?;
```

Comment:
I think this is/was a bug.

We're there any circumstances that caused `env` to be `Some`?
If `env` is not specified, we should use the environment that is in the lockfile.

This now lets us modify `LocalBackend::upload_artifact_dir` so that it uploads directly to the location that the `CachedBackend` expects, and bypass `get_tarball_url` and `raw_download`.

Reply:
We would have subjobs that would:

```sh
lal build
lal -e xenial publish {name}
```

and then if all the subjobs succeeded we'd re-publish one of them to the default bucket:

```sh
lal env reset
lal -e $(lal env) export {component}=${{BUILD_NUMBER}}
echo "Republishing $(lal env) component to default"
mkdir -p ARTIFACT
tar xzf {component}.tar.gz lockfile.json
mv {component}.tar.gz lockfile.json ARTIFACT/
lal publish {component}
```

Reply:
Or if we only had one env, we would sometimes just republish straight from the subjob.

```sh
+ lal -e xenial publish freetype
lal::publish: Publishing freetype=202 to xenial
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/env/xenial/freetype/202/freetype.tar.gz
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/env/xenial/freetype/202/freetype.tar.gz (X-Checksum-Sha1)
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/env/xenial/freetype/202/lockfile.json
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/env/xenial/freetype/202/lockfile.json (X-Checksum-Sha1)
+ lal publish freetype
lal::publish: Publishing freetype=202 to global
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/freetype/202/freetype.tar.gz
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/freetype/202/freetype.tar.gz (X-Checksum-Sha1)
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/freetype/202/lockfile.json
lal::storage::artifactory: PUT https://engci-maven.cisco.com/artifactory/CME-release/freetype/202/lockfile.json (X-Checksum-Sha1)
```

At any rate it's important to publish to the specific bucket because `fetch` is done from the specific env bucket in the environment we are in (unless overridden), but the global one is treated as the join. It's a bit messy atm. Would benefit from a rethink.